persist: refactor batch builder: merge user + compaction write paths

pH14 opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

The logic for writing user batches and compaction batches has diverged a bit, and this has led to a few rough edges. This PR merges their write path logic to both use `BatchBuilder` (and removes `ColumnarRecordsVecBuilder`), mostly moving the more up-to-date logic of compaction into `BatchBuilder`. Doing so helps us tackle a bunch of issues:

* We now correctly calculate part size for user batches (currently it's possible to write batches 100s of MiB over the target 128MiB depending on the key/value sizes)
* We calculate runs on user batches
* User batches are now sorted / consolidated
* Compaction code is much simplified

Additionally, having the same write path makes future MFP-related changes (like sorting updates by in-mem Key and Val `Ord` implementations) much simpler.

### Motivation

Fixes a few bugs on our M3 checklist:

* Calculate part size correctly for user records
* Consolidate user parts before writing
* persist: blobs in S3 larger than expected #14579

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

pH14 commented:
@danhhz this is still very rough, but I wouldn't mind a glance to make sure the overall direction here seems reasonable

danhhz commented on src/persist/src/indexed/columnar.rs:
```diff
@@ -143,24 +143,7 @@ where
-            builder.push(((key, val), Codec64::encode(ts), Codec64::encode(diff)))
-        }
-        ColumnarRecordsVec(builder.finish())
+        ColumnarRecordsVec(vec![])
```
WIP?

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -187,17 +187,26 @@ where
     blob: Arc<dyn Blob + Send + Sync>,
     metrics: Arc<Metrics>,
 
+    current_part: Vec<((Vec<u8>, Vec<u8>), T, D)>,
```
I mentioned this in the slack thread, but all these little allocations are bad for perf. if we're careful about lifetimes, should be possible to make all of those into slices. so have one big `Vec<u8>` that is cleared between parts and then then we collect things as `Vec<(&[u8], &[u8], T, D)>` or `BTreeMap<(&[u8], &[u8], T), D>`

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -187,17 +187,26 @@ where
+    current_part: Vec<((Vec<u8>, Vec<u8>), T, D)>,
+    current_part_total_bytes: usize,
+    current_part_key_bytes: usize,
+    current_part_value_bytes: usize,
```
feels like BatchBuilder has gotten pretty hefty folding all this in. I think there's an opportunity for something like `ColumnarRecordsBuilder` but instead of buffering it all directly into an in-progress ColumnarRecords, we do what you have here and buffer it first into something we can consolidate and then copy it into a ColumnarRecords

pH14 replied:
hrm I played around with this a bit, and I couldn't find a satisfying abstraction. I had some trouble following the existing code of `BatchBuilder -> ColumnarRecordsVecBuilder -> ColumnarRecordsBuilder -> ColumnarRecords` so I was hoping merging the first two steps together would be helpful, but it does lead to the sprawl of fields you see here...

danhhz replied:
combined with my comment about not doing a hard cap, I was thinking something like this (sorry for the huge diff!)

```diff
diff --git a/src/persist-client/src/batch.rs b/src/persist-client/src/batch.rs
index 34ed9b8f6..5f61c2da4 100644
--- a/src/persist-client/src/batch.rs
+++ b/src/persist-client/src/batch.rs
@@ -191,20 +191,12 @@ where
 {
     lower: Antichain<T>,
     max_ts: T,
-    blob_target_size: usize,
 
     shard_id: ShardId,
     blob: Arc<dyn Blob + Send + Sync>,
     metrics: Arc<Metrics>,
-    batch_write_metrics: BatchWriteMetrics,
-
-    key_buf: Vec<u8>,
-    val_buf: Vec<u8>,
 
-    current_part: Vec<((Range<usize>, Range<usize>), T, D)>,
-    current_part_total_bytes: usize,
-    current_part_key_bytes: usize,
-    current_part_value_bytes: usize,
+    buffer: BatchBuffer<T, D>,
 
     greatest_kv_seen: Option<(Vec<u8>, Vec<u8>)>,
     runs: Vec<usize>,
@@ -253,14 +245,13 @@ where
         Self {
             lower,
             max_ts: T::minimum(),
-            blob_target_size: cfg.blob_target_size,
             blob,
+            buffer: BatchBuffer::new(
+                Arc::clone(&metrics),
+                batch_write_metrics,
+                cfg.blob_target_size,
+            ),
             metrics,
-            batch_write_metrics,
-            current_part: Vec::new(),
-            current_part_total_bytes: 0,
-            current_part_key_bytes: 0,
-            current_part_value_bytes: 0,
             greatest_kv_seen: None,
             parts_written: 0,
             runs: Vec::new(),
@@ -268,8 +259,6 @@ where
             parts,
             shard_id,
             since,
-            key_buf: Vec::new(),
-            val_buf: Vec::new(),
             // TODO: The default case would ideally be `[self.max_ts+1]` but
             // there's nothing that lets us increment a timestamp. An empty
             // antichain is guaranteed to correctly bound the data in this
@@ -309,7 +298,8 @@ where
             });
         }
 
-        self.flush_current_part().await;
+        let columnar = self.buffer.drain();
+        self.flush_part(columnar).await;
 
         let parts = self.parts.finish().await;
 
@@ -348,6 +338,103 @@ where
 
         self.max_ts.join_assign(ts);
 
+        match self.buffer.push(key, val, ts.clone(), diff.clone()) {
+            Some(part_to_flush) => {
+                self.flush_part(part_to_flush).await;
+                Ok(Added::RecordAndParts)
+            }
+            None => Ok(Added::Record),
+        }
+    }
+
+    /// Flushes the current part to Blob storage, first consolidating and then columnar encoding
+    /// the updates. It is the caller's responsibility to chunk `current_part` to be no greater
+    /// than [crate::PersistConfig::blob_target_size], and must absolutely be less than
+    /// [mz_persist::indexed::columnar::KEY_VAL_DATA_MAX_LEN]
+    async fn flush_part(&mut self, columnar: ColumnarRecords) {
+        let num_updates = columnar.len();
+        if num_updates == 0 {
+            return;
+        }
+
+        match (
+            &self.greatest_kv_seen,
+            columnar.get(num_updates.saturating_sub(1)),
+        ) {
+            // our updates contain a key that exists within the range of a run we've
+            // already created. we should start a new run, as this part is no longer
+            // contiguous with the previous run/part
+            (Some(_greatest_kv_seen @ (k, v)), Some(greatest_kv_in_part))
+                if (k.as_slice(), v.as_slice()) > greatest_kv_in_part.0 =>
+            {
+                self.runs.push(self.parts_written);
+            }
+            (_, Some(_greatest_kv_in_batch @ ((k, v), _t, _d))) => {
+                self.greatest_kv_seen = Some((k.to_vec(), v.to_vec()));
+            }
+            (Some(_), None) | (None, None) => {}
+        };
+
+        let start = Instant::now();
+        self.parts
+            .write(columnar, self.inline_upper.clone(), self.since.clone())
+            .await;
+        self.metrics
+            .compaction
+            .batch
+            .step_part_writing
+            .inc_by(start.elapsed().as_secs_f64());
+
+        self.parts_written += 1;
+        self.num_updates += num_updates;
+    }
+}
+
+#[derive(Debug)]
+struct BatchBuffer<T, D> {
+    metrics: Arc<Metrics>,
+    batch_write_metrics: BatchWriteMetrics,
+    blob_target_size: usize,
+
+    key_buf: Vec<u8>,
+    val_buf: Vec<u8>,
+
+    current_part: Vec<((Range<usize>, Range<usize>), T, D)>,
+    current_part_total_bytes: usize,
+    current_part_key_bytes: usize,
+    current_part_value_bytes: usize,
+}
+
+impl<T, D> BatchBuffer<T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    fn new(
+        metrics: Arc<Metrics>,
+        batch_write_metrics: BatchWriteMetrics,
+        blob_target_size: usize,
+    ) -> Self {
+        BatchBuffer {
+            metrics,
+            batch_write_metrics,
+            blob_target_size,
+            key_buf: Default::default(),
+            val_buf: Default::default(),
+            current_part: Default::default(),
+            current_part_total_bytes: Default::default(),
+            current_part_key_bytes: Default::default(),
+            current_part_value_bytes: Default::default(),
+        }
+    }
+
+    fn push<K: Codec, V: Codec>(
+        &mut self,
+        key: &K,
+        val: &V,
+        ts: T,
+        diff: D,
+    ) -> Option<ColumnarRecords> {
         let initial_key_buf_len = self.key_buf.len();
         let initial_val_buf_len = self.val_buf.len();
         self.metrics
@@ -367,52 +454,26 @@ where
             let v = &self.val_buf[v_range.clone()];
             ColumnarRecordsBuilder::columnar_record_size(k, v)
         };
-        let (ts, diff) = (ts.clone(), diff.clone());
 
-        let mut part_written = false;
+        self.current_part_total_bytes += size;
+        self.current_part_key_bytes += k_range.len();
+        self.current_part_value_bytes += v_range.len();
+        self.current_part.push(((k_range, v_range), ts, diff));
+
         // if we've filled up a batch part, flush out to blob to keep our memory usage capped.
-        if size + self.current_part_total_bytes > self.blob_target_size {
+        if self.current_part_total_bytes > self.blob_target_size {
             // TODO: we currently consolidate a part when we've reached our target size of
             // _unconsolidated_ updates. This means the part we actually write to Blob may
             // be (substantially) smaller post-consolidation. We could alternatively create
             // a `BTreeMap<(&[u8], &[u8], T), D>`, consolidate as we go, and only write out
             // a part once the data referenced by the map reaches our target size.
-            self.flush_current_part().await;
-            part_written = true;
-        }
-
-        self.current_part_total_bytes += size;
-        self.current_part_key_bytes += k_range.len();
-        self.current_part_value_bytes += v_range.len();
-
-        if part_written {
-            // NB: we flush our shared buf's contents to Blob storage when we've hit our
-            // target size, but we can only determine whether we've hit the target size
-            // after encoding into the buf. this means that after flushing to Blob, we have
-            // one dangling update that we need to copy back after the buf is reset.
-            let dangling_key = self.key_buf[k_range].to_owned();
-            let dangling_val = self.val_buf[v_range].to_owned();
-
-            self.key_buf.clear();
-            self.val_buf.clear();
-
-            self.key_buf.extend_from_slice(&dangling_key);
-            self.val_buf.extend_from_slice(&dangling_val);
-
-            self.current_part
-                .push(((0..dangling_key.len(), 0..dangling_val.len()), ts, diff));
-            Ok(Added::RecordAndParts)
+            Some(self.drain())
         } else {
-            self.current_part.push(((k_range, v_range), ts, diff));
-            Ok(Added::Record)
+            None
         }
     }
 
-    /// Flushes the current part to Blob storage, first consolidating and then columnar encoding
-    /// the updates. It is the caller's responsibility to chunk `current_part` to be no greater
-    /// than [crate::PersistConfig::blob_target_size], and must absolutely be less than
-    /// [mz_persist::indexed::columnar::KEY_VAL_DATA_MAX_LEN]
-    async fn flush_current_part(&mut self) {
+    fn drain(&mut self) -> ColumnarRecords {
         let mut updates = Vec::with_capacity(self.current_part.len());
         for ((k_range, v_range), t, d) in self.current_part.drain(..) {
             updates.push(((&self.key_buf[k_range], &self.val_buf[v_range]), t, d));
@@ -425,26 +486,8 @@ where
             .inc_by(start.elapsed().as_secs_f64());
 
         if updates.is_empty() {
-            return;
+            return ColumnarRecordsBuilder::default().finish();
         }
-
-        self.num_updates += updates.len();
-
-        match (&self.greatest_kv_seen, updates.last()) {
-            // our updates contain a key that exists within the range of a run we've
-            // already created. we should start a new run, as this part is no longer
-            // contiguous with the previous run/part
-            (Some(_greatest_kv_seen @ (k, v)), Some(greatest_kv_in_part))
-                if (k.as_slice(), v.as_slice()) > greatest_kv_in_part.0 =>
-            {
-                self.runs.push(self.parts_written);
-            }
-            (_, Some(_greatest_kv_in_batch @ ((k, v), _t, _d))) => {
-                self.greatest_kv_seen = Some((k.to_vec(), v.to_vec()));
-            }
-            (Some(_), None) | (None, None) => {}
-        };
-
         let start = Instant::now();
         let mut builder = ColumnarRecordsBuilder::default();
         builder.reserve_exact(
@@ -463,21 +506,12 @@ where
             .step_columnar_encoding
             .inc_by(start.elapsed().as_secs_f64());
 
-        let start = Instant::now();
-        self.parts
-            .write(columnar, self.inline_upper.clone(), self.since.clone())
-            .await;
-        self.metrics
-            .compaction
-            .batch
-            .step_part_writing
-            .inc_by(start.elapsed().as_secs_f64());
-
-        self.parts_written += 1;
         self.current_part_total_bytes = 0;
         self.current_part_key_bytes = 0;
         self.current_part_value_bytes = 0;
         assert_eq!(self.current_part.len(), 0);
+
+        columnar
     }
 }
 
diff --git a/src/persist/src/indexed/columnar.rs b/src/persist/src/indexed/columnar.rs
index 0a62eef89..09e7d7d27 100644
--- a/src/persist/src/indexed/columnar.rs
+++ b/src/persist/src/indexed/columnar.rs
@@ -463,18 +463,6 @@ impl ColumnarRecordsBuilder {
     }
 }
 
-/// A new-type so we can impl FromIterator for Vec<ColumnarRecords>.
-#[derive(Debug)]
-pub struct ColumnarRecordsVec(pub Vec<ColumnarRecords>);
-
-impl ColumnarRecordsVec {
-    /// Unwraps this ColumnarRecordsVec, returning the underlying
-    /// Vec<ColumnarRecords>.
-    pub fn into_inner(self) -> Vec<ColumnarRecords> {
-        self.0
-    }
-}
-
 #[cfg(test)]
 mod tests {
     use super::*;
````

pH14 replied:
yeah that works! updated with largely this change

danhhz reviewed (commented):
seems sane! left a couple comments

also the `inline_upper: Option<Antichain<T>>` stuff feels kinda fishy, but I didn't load it all up again, so I'm probably just forgetting some inherent complexity

pH14 commented:
> also the inline_upper: Option<Antichain<T>> stuff feels kinda fishy, but I didn't load it all up again, so I'm probably just forgetting some inherent complexity

It is a little fishy for sure. Currently we write an inline upper of `[]` for all user batches, and then register them with a non-empty frontier. Compaction writes the same  non-empty upper for both inline and registered descriptions. The `Option<Antichain<T>>` is a way of differentiating those cases -- if a caller knows exactly what to use in its inline description it can pass one in, otherwise a `None` will default to an upper of `[]`

pH14 commented on src/persist-client/src/batch.rs:
```diff
@@ -254,43 +288,40 @@ where
+        // produced by compaction, which is allowed to produce batches with an arbitrary
+        // `since`)
+        if registered_upper.less_equal(&self.max_ts)
+            && !PartialOrder::less_equal(&registered_upper, &self.since)
```
this section feels a bit icky to me. not sure if there's a simpler expression of the logic, or a better way to phrase the comment

pH14 commented on src/persist-client/src/internal/metrics.rs:
```diff
@@ -571,6 +571,10 @@ pub struct BatchWriteMetrics {
+
+    pub(crate) step_consolidation: Counter,
+    pub(crate) step_columnar_encoding: Counter,
+    pub(crate) step_part_writing: Counter,
```
don't love adding the metrics here compared to `CompactionStepTimings` (which I find easier to use with how they're parameterized), but I didn't want to burn too many cycles on refactoring the metrics structure and overcomplicating this PR

pH14 commented:
I think this is ready for (re)review! I couldn't find the right way to simplify the `BatchBuilder` logic and how it handles the current part, so open to suggestions there, as well as a few spots I've highlighted. The code doesn't read as smoothly as I'd like yet, but it is functional and I'm going a little bleary-eyed staring at it further. Hopefully y'all are able to see what I cannot! 🕵️

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -328,47 +359,126 @@ where
+        self.current_part_value_bytes += v_range.len();
+
         if part_written {
+            // NB: we flush our shared buf's contents to Blob storage when we've hit our
```
this is to let us use `self.blob_target_size` as a hard cap? I don't think that's worth the complexity over unconditionally adding the update and flushing if the new size is >= self.blob_target_size the size of individual updates should be small compared to self.blob_target_size and also, there's all sorts of overhead in our columnar format that isn't taken into account in the `ColumnarRecordsBuilder::columnar_record_size` computation

pH14 replied:
okayokay fine, I'll accept the limit as a soft cap 😄

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -328,47 +359,126 @@ where
+            // already created. we should start a new run, as this part is no longer
+            // contiguous with the previous run/part
+            (Some(_greatest_kv_seen @ (k, v)), Some(greatest_kv_in_part))
+                if (k.as_slice(), v.as_slice()) > greatest_kv_in_part.0 =>
```
I think we actually want to compare by all of `(K, V, T)`

pH14 replied:
hmm why would we want to compare `T`? haven't been able to convince myself we should split off a new run in case a part has the same `(K, V)` but a lesser `T` to the current run

danhhz replied:
I was thinking of the invariants as (1) the whole batch is sorted by (K, V, T) and (2) that parts are non-overlapping. combined, I think that results in the derived requirement that parts care about (K, V, T). but, I think it's an interesting question about whether we need to sort by (K, V, T) over (K, V). I can't immediately think of anything

danhhz replied:
I just double checked and dd does order by (K, V, T) fwiw https://github.com/TimelyDataflow/differential-dataflow/blob/c2e8fefce9ddad0aef5afcac0238b4dc6ae0ddcb/src/trace/implementations/ord.rs#L100-L101

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -328,47 +359,126 @@ where
+                self.runs.push(self.parts_written);
+            }
+            (_, Some(_greatest_kv_in_batch @ ((k, v), _t, _d))) => {
+                self.greatest_kv_seen = Some((k.to_vec(), v.to_vec()));
```
we should reuse the existing `Vec<u8>` allocations if we can (i.e. a `clear` + `extend_from_slice`)

pH14 replied:
I'm less worried about this because it's only called per part, rather than per update, but I don't think readability took much of a hit, so.... done!

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -328,47 +359,126 @@ where
+
+        self.num_updates += updates.len();
+
+        match (&self.greatest_kv_seen, updates.last()) {
```
it took me quite a while to convince myself how this worked. maybe if we pull out an outer if let for greatest_kv_in_part?

```
if let Some(greatest_kv_in_part) = greatest_kv_in_part {
    match self.greatest_kv_seen.as_ref() {
        ...
    }
}
```

pH14 replied:
given the bug fixes, this whole section has been very rewritten and I think reads a lot more easily than before. lmk what you think

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -328,47 +359,126 @@ where
+            // our updates contain a key that exists within the range of a run we've
+            // already created. we should start a new run, as this part is no longer
+            // contiguous with the previous run/part
+            (Some(_greatest_kv_seen @ (k, v)), Some(greatest_kv_in_part))
```
hold up, shouldn't this be the least kv in the part? like if we emit `[a, c] and then [b, d]` it need to be two runs right?

pH14 replied:
Ooooooooo great spot! this has been a bug since the original bounded mem compaction PR

pH14 replied:
Oops and I think there's another bug in the run tracking. Currently we compare against the greatest KV seen, but really it should be of the most recent run... otherwise we might generate a new run rather than extending an existing

danhhz reviewed (commented):
I'm so excited to have this!

pH14 commented:
Getting close! Addressed the comments as much as I could and fixed a few bugs in how we tracked runs. I'll self-review again tomorrow morning and then I think we're ready for the last round (🤞)

danhhz commented on src/persist-client/tests/machine/batch:
```diff
@@ -0,0 +1,240 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
```
omg these tests are so good 😍

danhhz commented on src/persist-client/tests/machine/batch:
```diff
@@ -0,0 +1,240 @@
+----
+error: maximum timestamp 3 is beyond the expected batch upper: Antichain { elements: [2] }
+
+# but batches with since > upper are allowed to have updates with timestamps beyond upper
```
could also test that we error if the ts is gt both since AND upper

danhhz replied:
thought about this more. because upper is a strict upper bound on the "original" ts and the ts has been advanced_by since, then I think the actual hard invariant if upper <= since is `upper.advance_by(since) !< ts` (see [advance_by](https://docs.rs/differential-dataflow/latest/differential_dataflow/lattice/trait.Lattice.html#method.advance_by))

pH14 replied:
hm still wrapping my head around `advance_by`... one quick practical question: `advance_by` takes a `T` and advances it by `Antichain<T>`, but `upper` and `since` are both `Antichain<T>`, so I'm not sure what the comparable operation would be

pH14 replied:
for now, opted to mirror the logic we have in `encoding` and added the test you mentioned

danhhz replied:
> one quick practical question: advance_by takes a T and advances it by Antichain<T>, but upper and since are both Antichain<T>, so I'm not sure what the comparable operation would be

ah sorry, just saw this. advance_by is defined on Lattice, which is also implemented for Antichain<T: Lattice>

> for now, opted to mirror the logic we have in encoding and added the test you mentioned

wfm!

danhhz reviewed (approved):
\o/
